File: components/monitoring-v2/sdk/r/monitoring_sdk.R
# MonitoringSDK - R
# Lightweight TCP-based monitoring instrumentation
# Version: 2.0.0

library(R6)

#' MonitoringSDK Class
#' 
#' @description
#' Lightweight monitoring SDK for R using TCP socket communication
#' 
#' @details
#' Features:
#' - TCP socket communication
#' - Local buffering (1000 messages)
#' - Circuit breaker pattern
#' - Automatic reconnection
#' - <1ms overhead per call
#' 
#' @examples
#' \dontrun{
#' sdk <- MonitoringSDK$new(source = "my-r-service")
#' sdk$log_event("info", "Service started")
#' sdk$log_metric("requests_total", 42, "count")
#' sdk$close()
#' }
#' 
#' @export
MonitoringSDK <- R6Class("MonitoringSDK",
  public = list(
    #' @field source Source identifier
    source = NULL,
    
    #' @field tcp_host TCP host
    tcp_host = NULL,
    
    #' @field tcp_port TCP port
    tcp_port = NULL,
    
    #' @field debug Debug mode
    debug = FALSE,
    
    #' @description Create a new MonitoringSDK instance
    #' @param source Source identifier (service/script name)
    #' @param tcp_host TCP host (default: "localhost")
    #' @param tcp_port TCP port (default: 17000)
    #' @param timeout Connection timeout in seconds (default: 5)
    #' @param debug Enable debug logging (default: FALSE)
    initialize = function(source, tcp_host = "localhost", tcp_port = 17000, 
                         timeout = 5, debug = FALSE) {
      if (missing(source) || is.null(source) || source == "") {
        stop("source is required")
      }
      
      self$source <- source
      self$tcp_host <- tcp_host
      self$tcp_port <- tcp_port
      self$debug <- debug
      
      private$.timeout <- timeout
      private$.state <- "disconnected"
      private$.socket <- NULL
      private$.buffer <- list()
      private$.overflow_count <- 0
      private$.reconnect_delay <- 1.0
      private$.last_reconnect <- 0
      
      # Context
      private$.trace_id <- NULL
      private$.span_id <- NULL
      private$.context <- list()
      
      # Statistics
      private$.messages_sent <- 0
      private$.messages_buffered <- 0
      private$.messages_dropped <- 0
      private$.reconnect_count <- 0
      
      # Initial connection
      private$connect()
    },
    
    #' @description Log an event
    #' @param level Log level (debug, info, warn, error, fatal)
    #' @param message Event message
    #' @param context Optional context list
    log_event = function(level, message, context = list()) {
      private$send_message(list(
        type = "event",
        data = list(
          level = level,
          msg = message,
          ctx = context
        )
      ))
    },
    
    #' @description Log a metric
    #' @param name Metric name
    #' @param value Metric value
    #' @param unit Unit of measurement
    #' @param tags Optional tags list
    log_metric = function(name, value, unit = "", tags = list()) {
      private$send_message(list(
        type = "metric",
        data = list(
          name = name,
          value = as.numeric(value),
          unit = unit,
          tags = tags
        )
      ))
    },
    
    #' @description Log job progress
    #' @param job_id Job identifier
    #' @param percent Completion percentage (0-100)
    #' @param status Status description
    log_progress = function(job_id, percent, status = "running") {
      percent <- max(0, min(100, as.integer(percent)))
      
      private$send_message(list(
        type = "progress",
        data = list(
          job_id = job_id,
          percent = percent,
          status = status
        )
      ))
    },
    
    #' @description Log resource usage
    #' @param cpu_percent CPU usage percentage (NULL for auto)
    #' @param memory_mb Memory usage in MB (NULL for auto)
    #' @param disk_io_mb Disk I/O in MB (NULL for auto)
    #' @param network_io_mb Network I/O in MB (NULL for auto)
    log_resource = function(cpu_percent = NULL, memory_mb = NULL, disk_io_mb = NULL, network_io_mb = NULL) {
      # Auto-collect metrics if not provided
      if (is.null(cpu_percent))    cpu_percent <- private$get_cpu_percent()
      if (is.null(memory_mb))      memory_mb <- private$get_memory_mb()
      if (is.null(disk_io_mb))     disk_io_mb <- private$get_disk_io_mb()
      if (is.null(network_io_mb))  network_io_mb <- private$get_network_io_mb()
      
      private$send_message(list(
        type = "resource",
        data = list(
          cpu = as.numeric(cpu_percent),
          mem = as.numeric(memory_mb),
          disk = as.numeric(disk_io_mb),
          net = as.numeric(network_io_mb),
          pid = Sys.getpid()
        )
      ))
    },
    
    #' @description Start a distributed trace span
    #' @param name Span name
    #' @param trace_id Optional trace ID (NULL to generate)
    #' @return Span ID
    start_span = function(name, trace_id = NULL) {
      # Generate span ID
      span_id <- private$generate_id()
      
      # Set trace ID if provided
      if (!is.null(trace_id)) {
        private$.trace_id <- trace_id
      }
      
      # Generate trace ID if not set
      if (is.null(private$.trace_id)) {
        private$.trace_id <- private$generate_id()
      }
      
      # Store current span as parent
      parent_span_id <- private$.span_id
      
      # Set new span as current
      private$.span_id <- span_id
      
      # Send span start event
      msg <- list(
        type = "span",
        tid = private$.trace_id,
        sid = span_id,
        data = list(
          name = name,
          start = as.numeric(Sys.time()) * 1000,
          end = NULL,
          status = "started",
          tags = list()
        )
      )
      
      if (!is.null(parent_span_id)) {
        msg$pid <- parent_span_id
      }
      
      private$send_message(msg)
      
      return(span_id)
    },
    
    #' @description End a distributed trace span
    #' @param span_id Span ID from start_span
    #' @param status Span status (default: "success")
    #' @param tags Optional tags list
    end_span = function(span_id, status = "success", tags = list()) {
      private$send_message(list(
        type = "span",
        tid = private$.trace_id,
        sid = span_id,
        data = list(
          name = "",
          start = 0,
          end = as.numeric(Sys.time()) * 1000,
          status = status,
          tags = tags
        )
      ))
      
      # Clear current span if it matches
      if (!is.null(private$.span_id) && private$.span_id == span_id) {
        private$.span_id <- NULL
      }
    },
    
    #' @description Set trace ID for correlation
    #' @param trace_id Trace ID
    set_trace_id = function(trace_id) {
      private$.trace_id <- trace_id
    },
    
    #' @description Set context variable
    #' @param key Context key
    #' @param value Context value
    set_context = function(key, value) {
      private$.context[[key]] <- value
    },
    
    #' @description Get SDK statistics
    #' @return List of statistics
    get_stats = function() {
      list(
        source = self$source,
        state = private$.state,
        messages_sent = private$.messages_sent,
        messages_buffered = private$.messages_buffered,
        messages_dropped = private$.messages_dropped,
        buffer_size = length(private$.buffer),
        overflow_count = private$.overflow_count,
        reconnect_count = private$.reconnect_count
      )
    },
    
    #' @description Close connection and cleanup
    close = function() {
      # Send goodbye message
      if (private$.state == "connected" && !is.null(private$.socket)) {
        tryCatch({
          private$send_message(list(type = "goodbye", data = list()))
        }, error = function(e) {})
      }
      
      # Close socket
      if (!is.null(private$.socket)) {
        tryCatch({
          close(private$.socket)
        }, error = function(e) {})
        private$.socket <- NULL
      }
      
      private$.state <- "disconnected"
      
      if (self$debug) {
        message(sprintf("[MonitoringSDK] Closed. Messages sent: %d, buffered: %d, dropped: %d",
                       private$.messages_sent, private$.messages_buffered, private$.messages_dropped))
      }
    }
  ),
  
  private = list(
    .timeout = NULL,
    .state = NULL,
    .socket = NULL,
    .buffer = NULL,
    .overflow_count = NULL,
    .reconnect_delay = NULL,
    .last_reconnect = NULL,
    
    .trace_id = NULL,
    .span_id = NULL,
    .context = NULL,
    
    .messages_sent = NULL,
    .messages_buffered = NULL,
    .messages_dropped = NULL,
    .reconnect_count = NULL,
    
    MAX_BUFFER_SIZE = 1000,
    MAX_MESSAGE_SIZE = 65536,
    MAX_RECONNECT_DELAY = 30,
    
    connect = function() {
      # Already connected
      if (private$.state == "connected" && !is.null(private$.socket)) {
        return(TRUE)
      }
      
      # Throttle reconnection attempts
      now <- as.numeric(Sys.time())
      if (now - private$.last_reconnect < private$.reconnect_delay) {
        return(FALSE)
      }
      private$.last_reconnect <- now
      
      tryCatch({
        private$.socket <- socketConnection(
          host = self$tcp_host,
          port = self$tcp_port,
          blocking = FALSE,
          open = "w+b",
          timeout = private$.timeout
        )
        
        private$.state <- "connected"
        private$.reconnect_delay <- 1.0
        private$.reconnect_count <- private$.reconnect_count + 1
        
        if (self$debug) {
          message(sprintf("[MonitoringSDK] Connected to %s:%d", self$tcp_host, self$tcp_port))
        }
        
        # Flush buffer
        private$flush_buffer()
        
        return(TRUE)
        
      }, error = function(e) {
        private$.state <- "disconnected"
        
        if (self$debug) {
          message(sprintf("[MonitoringSDK] Connection failed: %s", e$message))
        }
        
        # Exponential backoff
        private$.reconnect_delay <- min(private$.reconnect_delay * 2, private$MAX_RECONNECT_DELAY)
        
        if (!is.null(private$.socket)) {
          tryCatch(close(private$.socket), error = function(e) {})
          private$.socket <- NULL
        }
        
        return(FALSE)
      })
    },
    
    send_message = function(msg) {
      # Add protocol fields
      msg$v <- 1
      msg$src <- self$source
      msg$ts <- as.numeric(Sys.time()) * 1000  # Unix millis
      
      # Add trace context if available
      if (!is.null(private$.trace_id)) {
        msg$tid <- private$.trace_id
      }
      if (!is.null(private$.span_id)) {
        msg$sid <- private$.span_id
      }
      
      # Serialize to JSON
      json_msg <- tryCatch({
        jsonlite::toJSON(msg, auto_unbox = TRUE)
      }, error = function(e) {
        if (self$debug) {
          message(sprintf("[MonitoringSDK] Serialization error: %s", e$message))
        }
        private$.messages_dropped <- private$.messages_dropped + 1
        return(NULL)
      })
      
      if (is.null(json_msg)) {
        return(FALSE)
      }
      
      message_str <- paste0(json_msg, "\n")
      
      # Check message size
      if (nchar(message_str) > private$MAX_MESSAGE_SIZE) {
        if (self$debug) {
          message(sprintf("[MonitoringSDK] Message too large: %d bytes", nchar(message_str)))
        }
        private$.messages_dropped <- private$.messages_dropped + 1
        return(FALSE)
      }
      
      # Try to send
      if (private$.state == "connected" && !is.null(private$.socket)) {
        tryCatch({
          writeLines(message_str, private$.socket, sep = "")
          flush(private$.socket)
          private$.messages_sent <- private$.messages_sent + 1
          return(TRUE)
        }, error = function(e) {
          private$.state <- "disconnected"
          if (self$debug) {
            message(sprintf("[MonitoringSDK] Send failed: %s", e$message))
          }
          private$buffer_message(msg)
          private$connect()
          return(FALSE)
        })
      } else {
        # Not connected, buffer message
        private$buffer_message(msg)
        private$connect()
        return(FALSE)
      }
    },
    
    buffer_message = function(msg) {
      if (length(private$.buffer) < private$MAX_BUFFER_SIZE) {
        private$.buffer[[length(private$.buffer) + 1]] <- msg
        private$.messages_buffered <- private$.messages_buffered + 1
      } else {
        # Buffer overflow
        private$.state <- "overflow"
        private$.overflow_count <- private$.overflow_count + 1
        private$.messages_dropped <- private$.messages_dropped + 1
        
        if (private$.overflow_count %% 100 == 0 && self$debug) {
          message(sprintf("[MonitoringSDK] Buffer overflow! Dropped %d messages",
                         private$.overflow_count))
        }
      }
    },
    
    flush_buffer = function() {
      if (private$.state != "connected" || is.null(private$.socket)) {
        return()
      }
      
      flushed <- 0
      while (length(private$.buffer) > 0 && private$.state == "connected") {
        msg <- private$.buffer[[1]]
        private$.buffer[[1]] <- NULL
        
        if (private$send_message(msg)) {
          flushed <- flushed + 1
        } else {
          # Failed, put back
          private$.buffer <- c(list(msg), private$.buffer)
          break
        }
      }
      
      if (flushed > 0 && self$debug) {
        message(sprintf("[MonitoringSDK] Flushed %d buffered messages", flushed))
      }
      
      # Reset overflow state if buffer cleared
      if (length(private$.buffer) == 0 && private$.state == "overflow") {
        private$.state <- "connected"
        private$.overflow_count <- 0
      }
    },
    
    generate_id = function() {
      # Generate random 16-character ID
      paste0(sample(c(letters, LETTERS, 0:9), 16, replace = TRUE), collapse = "")
    },
    
    # Auto-collect CPU usage percentage
    get_cpu_percent = function() {
      tryCatch({
        if (file.exists("/proc/stat")) {
          lines <- readLines("/proc/stat", n = 1)
          vals <- as.numeric(strsplit(lines, "\\s+")[[1]][-1])
          if (length(vals) >= 4) {
            user <- vals[1]; nice <- vals[2]; system <- vals[3]; idle <- vals[4]
            total <- user + nice + system + idle
            used <- user + nice + system
            return(if (total > 0) (used / total) * 100 else 0)
          }
        }
        # Fallback: try system command
        cpu_info <- system("top -bn1 | grep 'Cpu(s)'", intern = TRUE)
        if (length(cpu_info) > 0) {
          # Parse CPU idle and calculate used
          idle <- as.numeric(sub(".*?([0-9.]+)%?.*id.*", "\\1", cpu_info[1]))
          return(100 - idle)
        }
        return(0)
      }, error = function(e) 0)
    },
    
    # Auto-collect memory usage in MB
    get_memory_mb = function() {
      tryCatch({
        if (file.exists("/proc/meminfo")) {
          lines <- readLines("/proc/meminfo")
          mem_total <- as.numeric(sub("MemTotal:\\s+(\\d+).*", "\\1", grep("MemTotal", lines, value = TRUE)))
          mem_available <- as.numeric(sub("MemAvailable:\\s+(\\d+).*", "\\1", grep("MemAvailable", lines, value = TRUE)))
          if (length(mem_total) > 0 && length(mem_available) > 0) {
            used_kb <- mem_total - mem_available
            return(used_kb / 1024)  # Convert to MB
          }
        }
        # Fallback: try free command
        mem_info <- system("free -m | grep Mem", intern = TRUE)
        if (length(mem_info) > 0) {
          vals <- as.numeric(strsplit(mem_info, "\\s+")[[1]])
          if (length(vals) >= 3) return(vals[3])  # Used memory
        }
        return(0)
      }, error = function(e) 0)
    },
    
    # Auto-collect disk I/O in MB
    get_disk_io_mb = function() {
      tryCatch({
        if (file.exists("/proc/diskstats")) {
          lines <- readLines("/proc/diskstats")
          total_sectors <- 0
          for (line in lines) {
            parts <- strsplit(trimws(line), "\\s+")[[1]]
            if (length(parts) >= 10) {
              read_sectors <- as.numeric(parts[6])
              write_sectors <- as.numeric(parts[10])
              total_sectors <- total_sectors + read_sectors + write_sectors
            }
          }
          # Convert sectors to MB (sector = 512 bytes)
          return((total_sectors * 512) / (1024 * 1024))
        }
        return(0)
      }, error = function(e) 0)
    },
    
    # Auto-collect network I/O in MB
    get_network_io_mb = function() {
      tryCatch({
        if (file.exists("/proc/net/dev")) {
          lines <- readLines("/proc/net/dev")
          total_bytes <- 0
          for (line in lines[-(1:2)]) {  # Skip header lines
            parts <- strsplit(trimws(line), "\\s+")[[1]]
            if (length(parts) >= 10) {
              # RX bytes (column 2) + TX bytes (column 10)
              rx_bytes <- as.numeric(sub(".*:", "", parts[1]))
              if (!is.na(rx_bytes)) {
                tx_bytes <- as.numeric(parts[9])
                total_bytes <- total_bytes + rx_bytes + tx_bytes
              }
            }
          }
          return(total_bytes / (1024 * 1024))  # Convert to MB
        }
        return(0)
      }, error = function(e) 0)
    }
  ),
  
  active = list(
    #' @field connected Check if connected
    connected = function() {
      private$.state == "connected"
    }
  )
)

# Example usage
if (FALSE) {
  # Create SDK
  sdk <- MonitoringSDK$new(source = "test-r-service", debug = TRUE)
  
  # Log events
  sdk$log_event("info", "Service started")
  sdk$log_event("error", "Connection failed", list(host = "db.example.com"))
  
  # Log metrics
  sdk$log_metric("requests_total", 42, "count")
  sdk$log_metric("response_time_ms", 125.5, "milliseconds")
  
  # Job progress
  sdk$log_progress("job-123", 50, "processing")
  
  # Resource usage
  sdk$log_resource(45.2, 2048, 100, 50)
  
  # Distributed tracing
  span_id <- sdk$start_span("process_request")
  Sys.sleep(0.1)
  sdk$end_span(span_id, "success")
  
  # Show stats
  print(sdk$get_stats())
  
  # Cleanup
  sdk$close()
}

File: components/monitoring-v2/sdk/python/monitoring_sdk.py
"""
MonitoringSDK - Python
Lightweight TCP-based monitoring instrumentation

Version: 2.0.0
"""

import socket
import json
import time
import threading
from enum import Enum
from typing import Optional, Dict, Any, List
from collections import deque
import random
import string


class State(Enum):
    """SDK connection states"""
    DISCONNECTED = "disconnected"
    CONNECTED = "connected"
    OVERFLOW = "overflow"


class MonitoringSDK:
    """
    Lightweight monitoring SDK for Python
    
    Features:
    - TCP socket communication
    - Local buffering (1000 messages)
    - Circuit breaker pattern
    - Automatic reconnection
    - <1ms overhead per call
    
    Example:
        sdk = MonitoringSDK(source='my-service')
        sdk.log_event('info', 'Service started')
        sdk.log_metric('requests_total', 42, 'count')
        sdk.close()
    """
    
    # Constants
    PROTOCOL_VERSION = 1
    MAX_MESSAGE_SIZE = 65536  # 64KB
    MAX_BUFFER_SIZE = 1000
    MAX_RECONNECT_DELAY = 30
    
    def __init__(self, source: str, tcp_host: str = 'localhost', 
                 tcp_port: int = 17000, timeout: float = 5.0, debug: bool = False):
        self.source = source
        self.tcp_host = tcp_host
        self.tcp_port = tcp_port
        self.timeout = timeout
        self.debug = debug
        
        # State
        self.state = State.DISCONNECTED
        self.socket: Optional[socket.socket] = None
        self.buffer: deque = deque(maxlen=self.MAX_BUFFER_SIZE)
        self.overflow_count = 0
        self.reconnect_delay = 1.0
        self.last_reconnect = 0
        
        # Context (for tracing)
        self.trace_id: Optional[str] = None
        self.span_id: Optional[str] = None
        self.context: Dict[str, Any] = {}
        
        # Statistics
        self.messages_sent = 0
        self.messages_buffered = 0
        self.messages_dropped = 0
        self.reconnect_count = 0
        
        # Thread safety
        self._lock = threading.RLock()
        
        # Initial connection
        self._connect()
    
    def _connect(self) -> bool:
        """Establish TCP connection"""
        with self._lock:
            # Already connected
            if self.state == State.CONNECTED and self.socket:
                return True
            
            # Throttle reconnection attempts
            now = time.time()
            if now - self.last_reconnect < self.reconnect_delay:
                return False
            
            self.last_reconnect = now
            
            try:
                self.socket = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
                self.socket.settimeout(self.timeout)
                self.socket.connect((self.tcp_host, self.tcp_port))
                
                self.state = State.CONNECTED
                self.reconnect_delay = 1.0  # Reset backoff
                self.reconnect_count += 1
                
                self._log(f"Connected to {self.tcp_host}:{self.tcp_port}")
                
                # Flush buffer
                self._flush_buffer()
                
                return True
            
            except Exception as e:
                self.state = State.DISCONNECTED
                self._log(f"Connection failed: {e}")
                
                # Exponential backoff
                self.reconnect_delay = min(self.reconnect_delay * 2, self.MAX_RECONNECT_DELAY)
                
                if self.socket:
                    try:
                        self.socket.close()
                    except:
                        pass
                    self.socket = None
                
                return False
    
    def _send_message(self, msg: Dict[str, Any]) -> bool:
        """Send message to sidecar"""
        with self._lock:
            # Add protocol fields
            msg['v'] = self.PROTOCOL_VERSION
            msg['src'] = self.source
            msg['ts'] = int(time.time() * 1000)  # Unix millis
            
            # Add trace context if available
            if self.trace_id:
                msg['tid'] = self.trace_id
            if self.span_id:
                msg['sid'] = self.span_id
            
            # Serialize to JSON
            try:
                json_msg = json.dumps(msg, separators=(',', ':'))
                message = (json_msg + '\n').encode('utf-8')
            except Exception as e:
                self._log(f"Serialization error: {e}")
                self.messages_dropped += 1
                return False
            
            # Check message size
            if len(message) > self.MAX_MESSAGE_SIZE:
                self._log(f"Message too large: {len(message)} bytes")
                self.messages_dropped += 1
                return False
            
            # Try to send
            if self.state == State.CONNECTED and self.socket:
                try:
                    self.socket.sendall(message)
                    self.messages_sent += 1
                    return True
                
                except Exception as e:
                    # Send failed, buffer message
                    self.state = State.DISCONNECTED
                    self._log(f"Send failed: {e}")
                    self._buffer_message(msg)
                    self._connect()  # Attempt reconnect
                    return False
            
            else:
                # Not connected, buffer message
                self._buffer_message(msg)
                self._connect()  # Attempt reconnect
                return False
    
    def _buffer_message(self, msg: Dict[str, Any]):
        """Buffer message locally"""
        if len(self.buffer) < self.MAX_BUFFER_SIZE:
            self.buffer.append(msg)
            self.messages_buffered += 1
        else:
            # Buffer overflow
            self.state = State.OVERFLOW
            self.overflow_count += 1
            self.messages_dropped += 1
            
            if self.overflow_count % 100 == 0:
                self._log(f"Buffer overflow! Dropped {self.overflow_count} messages")
    
    def _flush_buffer(self):
        """Flush buffered messages"""
        if self.state != State.CONNECTED or not self.socket:
            return
        
        flushed = 0
        while self.buffer and self.state == State.CONNECTED:
            msg = self.buffer.popleft()
            if self._send_message(msg):
                flushed += 1
            else:
                # Failed, put back
                self.buffer.appendleft(msg)
                break
        
        if flushed > 0:
            self._log(f"Flushed {flushed} buffered messages")
        
        # Reset overflow state if buffer cleared
        if not self.buffer and self.state == State.OVERFLOW:
            self.state = State.CONNECTED
            self.overflow_count = 0
    
    # Public API
    
    def log_event(self, level: str, message: str, context: Optional[Dict[str, Any]] = None) -> bool:
        """Log an event"""
        return self._send_message({
            'type': 'event',
            'data': {
                'level': level,
                'msg': message,
                'ctx': context or {}
            }
        })
    
    def log_metric(self, name: str, value: float, unit: str = '', 
                   tags: Optional[Dict[str, str]] = None) -> bool:
        """Log a metric"""
        return self._send_message({
            'type': 'metric',
            'data': {
                'name': name,
                'value': float(value),
                'unit': unit,
                'tags': tags or {}
            }
        })
    
    def log_progress(self, job_id: str, percent: int, status: str = 'running') -> bool:
        """Log job progress"""
        percent = max(0, min(100, int(percent)))
        return self._send_message({
            'type': 'progress',
            'data': {
                'job_id': job_id,
                'percent': percent,
                'status': status
            }
        })
    
    def log_resource(self, cpu_percent: float, memory_mb: float, 
                     disk_io_mb: float, network_io_mb: float) -> bool:
        """Log resource usage"""
        return self._send_message({
            'type': 'resource',
            'data': {
                'cpu': float(cpu_percent),
                'mem': float(memory_mb),
                'disk': float(disk_io_mb),
                'net': float(network_io_mb)
            }
        })
    
    def start_span(self, name: str, trace_id: Optional[str] = None) -> str:
        """Start a distributed trace span"""
        # Generate span ID
        span_id = self._generate_id()
        
        # Set trace ID if provided
        if trace_id:
            self.trace_id = trace_id
        
        # Generate trace ID if not set
        if not self.trace_id:
            self.trace_id = self._generate_id()
        
        # Store current span as parent
        parent_span_id = self.span_id
        
        # Set new span as current
        self.span_id = span_id
        
        # Send span start event
        msg = {
            'type': 'span',
            'tid': self.trace_id,
            'sid': span_id,
            'data': {
                'name': name,
                'start': int(time.time() * 1000),
                'end': None,
                'status': 'started',
                'tags': {}
            }
        }
        
        if parent_span_id:
            msg['pid'] = parent_span_id
        
        self._send_message(msg)
        return span_id
    
    def end_span(self, span_id: str, status: str = 'success', 
                 tags: Optional[Dict[str, Any]] = None) -> bool:
        """End a distributed trace span"""
        result = self._send_message({
            'type': 'span',
            'tid': self.trace_id,
            'sid': span_id,
            'data': {
                'name': '',
                'start': 0,
                'end': int(time.time() * 1000),
                'status': status,
                'tags': tags or {}
            }
        })
        
        # Clear current span if it matches
        if self.span_id == span_id:
            self.span_id = None
        
        return result
    
    def set_context(self, key: str, value: Any):
        """Set context variable"""
        self.context[key] = value
    
    def set_trace_id(self, trace_id: str):
        """Set trace ID for correlation"""
        self.trace_id = trace_id
    
    def close(self):
        """Close connection and cleanup"""
        with self._lock:
            # Send goodbye message
            if self.state == State.CONNECTED and self.socket:
                try:
                    self._send_message({'type': 'goodbye', 'data': {}})
                except:
                    pass
            
            # Close socket
            if self.socket:
                try:
                    self.socket.close()
                except:
                    pass
                self.socket = None
            
            self.state = State.DISCONNECTED
            
            self._log(f"Closed. Messages sent: {self.messages_sent}, "
                     f"buffered: {self.messages_buffered}, "
                     f"dropped: {self.messages_dropped}")
    
    def get_stats(self) -> Dict[str, Any]:
        """Get SDK statistics"""
        return {
            'source': self.source,
            'state': self.state.value,
            'messages_sent': self.messages_sent,
            'messages_buffered': self.messages_buffered,
            'messages_dropped': self.messages_dropped,
            'buffer_size': len(self.buffer),
            'overflow_count': self.overflow_count,
            'reconnect_count': self.reconnect_count
        }
    
    def _generate_id(self) -> str:
        """Generate random ID"""
        return ''.join(random.choices(string.ascii_letters + string.digits, k=16))
    
    def _log(self, message: str):
        """Internal logging"""
        if self.debug:
            print(f"[MonitoringSDK] {message}")
    
    def __enter__(self):
        """Context manager support"""
        return self
    
    def __exit__(self, exc_type, exc_val, exc_tb):
        """Context manager cleanup"""
        self.close()
        return False
    
    def __del__(self):
        """Destructor"""
        try:
            self.close()
        except:
            pass


# Example usage
if __name__ == '__main__':
    # Create SDK
    with MonitoringSDK(source='test-service', debug=True) as sdk:
        # Log events
        sdk.log_event('info', 'Service started')
        sdk.log_event('error', 'Connection failed', {'host': 'db.example.com'})
        
        # Log metrics
        sdk.log_metric('requests_total', 42, 'count')
        sdk.log_metric('response_time_ms', 125.5, 'milliseconds')
        
        # Job progress
        sdk.log_progress('job-123', 50, 'processing')
        
        # Resource usage
        sdk.log_resource(45.2, 2048, 100, 50)
        
        # Distributed tracing
        span_id = sdk.start_span('process_request')
        time.sleep(0.1)
        sdk.end_span(span_id, 'success')
        
        # Show stats
        print(f"Stats: {sdk.get_stats()}")

